curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java

```diff
@@ -770,14 +790,24 @@ static <T> void backgroundCreateParentsThenNode(final CuratorFrameworkImpl clien
             @Override
             public void performBackgroundOperation(OperationAndData<T> dummy) throws Exception
             {
+                AtomicReference<CuratorEvent> event = new AtomicReference<>();
+                AsyncCallback.Create2Callback callback = new AsyncCallback.Create2Callback()
+                {
+                    @Override
+                    public void processResult(int rc, String path, Object ctx, String name, Stat stat)
+                    {
+                        event.set(createResponseEvent(client, rc, path, ctx, name, stat));
+                    }
+                };
                 try
                 {
-                    ZKPaths.mkdirs(client.getZooKeeper(), path, false, aclProvider, createParentsAsContainers);
+                    ZKPaths.mkdirs(client.getZooKeeper(), path, false, aclProvider, createParentsAsContainers, callback);
                 }
                 catch ( KeeperException e )
                 {
                     if ( !client.getZookeeperClient().getRetryPolicy().allowRetry(e) )
                     {
+                        client.processBackgroundOperation(mainOperationAndData, event.get());
```

Comment:
Sounds like all we need in exceptional case is `rc` and `path`. `KeeperException` already covers these two `KeeperException::code` and `KeeperException::getPath`. So I think we could construct `CuratorEvent` without resorting to a callback.

Reply:
@kezhuw Thanks for your insight!

I checked that on exception, `stat` should be `null`, `ctx` we passed is `null`, and since we don't create sequential node here, `path == name`. So you're right that I can tidy the code a lot.

Reply:
Updated.

Reply:
@tisonkun On exception, `name` should be `null` as no node is created. But this won't affect final result as `KeeperException` needs only `CuratorEvent::getResultCode` and `CuratorEvent::getPath` to reconstruct itself.